SOFTWARE-3665 Move osg-flock document to the OSG Docs (#535)
opensciencegrid/docs: docs/submit/osg-flock.md

Configuring a submit host to flock to OSG
=========================================

> Overview

Reviewer: ```suggestion
```

> --------

Reviewer: ```suggestion
```

Requirements
------------

The requirements are:

* A public IP address, forward and reverse DNS.
* Ability to open a few firewall incoming ports to the WAN.
* HTCondor has to authenticate via pool GSI (preffered) or password. For GSI, the submit host
   has to have a host certificate.
* Reporting to the OSG accounting system has to be enabled. This can
   be accomplished by installing and configuring the *gratia-probe-condor* and *gratia-probe-glideinwms* RPMs.
* Submitted jobs should have the *+ProjectName* attribute specified with
   a valid registered project name.

Required Packages
-----------------

Enable the [OSG Yum repository](/common/yum/).

Install the packages required:

```console
root@host # yum install osg-flock
```

Gratia Probe Configuration
--------------------------

Edit the `/etc/gratia/condor/ProbeConfig` file. Change the probe section to be:

```file
    ProbeName="condor:YOUR_HOSTNAME_HERE"
    SiteName="YOUR_SITE_HERE"
    Grid="OSG"
    SuppressUnknownVORecords="0"
    SuppressNoDNRecords="0"
    SuppressGridLocalRecords="1"
    QuarantineUnknownVORecords="1"
    EnableProbe="1"

    MapUnknownToGroup="1"
    MapGroupToRole="1"
    VOOverride="osg"

> ```

Reviewer: Could we change everything above this line to two steps?

```
1. Copy over the recommended probe configuration:

    cp /etc/gratia/condor/Probeconfig-flocking /etc/gratia/condor/Probeconfig

1. Fill in the values for `ProbeName` and `SiteName` with the hostname and Topology resource name, respectively.
    For example:

        :::xml
        ProbeName="condor:foo.example.edu"
        SiteName="OSG_US_RESOURCE_SUBMIT"
```

Please remember to enable and start the probe:

```console
root@host # systemctl enable gratia-probes-cron
root@host # systemctl start gratia-probes-cron
```

GSI: HTCondor Configuration

> ---------------------------

Reviewer: @rynge since we ship this configuration in `osg-flock`, can we just get rid of this section?

Reviewer 2: I think we should. Maybe structure it as osg-flock is strongly recommended, but if that can be done, use pool passwords?

Reviewer: That makes sense. @efajardo, can you 

1. Get rid of this section
1. Change the pool password heading to indicate that it's optional
1. Write a sentence or two at the top of pool password section indicating that admins should only configure their schedd this way if they can't get a certificate
1. Tell them that they should replace the contents of `/etc/condor/config.d/80-osg-flocking.conf` with the config as described.

@rynge we could probably make this a little easier to config for admins: we can put the pool password config in an if block that triggers off the admin adding a config value like `OSG_FLOCKING_POOL_PASSWORD = True`

    SEC_CLIENT_INTEGRITY = OPTIONAL
    SEC_READ_ENCRYPTION = OPTIONAL
    SEC_CLIENT_ENCRYPTION = OPTIONAL
    
    #-- With strong security, do not use IP based controls
    HOSTALLOW_WRITE = *

    # OSG VO flocking host
    OSG_FLOCK = flock.opensciencegrid.org
    OSG_FLOCK_DN = /DC=org/DC=incommon/C=US/postalCode=53706/ST=WI/L=Madison/street=1210 West Dayton Street/O=University of Wisconsin-Madison/OU=OCIS/CN=flock.opensciencegrid.org

    # submitters/campus factories to flock to
    FLOCK_TO = $(FLOCK_TO) $(OSG_FLOCK)

    # Who to trust?  Inlucde the <VO>_DN to trust the submitter
    GSI_DAEMON_NAME = $(GSI_DAEMON_NAME), $(OSG_FLOCK_DN)

    # This host's certificates
    # These are needed if you are flocking to a glideinwms host
    GSI_DAEMON_CERT = /etc/grid-security/hostcert.pem
    GSI_DAEMON_KEY = /etc/grid-security/hostkey.pem
```

GSI: Requesting a Host Certificate
---------------------------------

A host certificate is used for authenticating your submit host to the OSG
infrastructure. If you do not already have a certificate, you can request one
using [these instructions](/security/host-certs/)

Pool Password: HTCondor Configuration

> -------------------------------------

Reviewer: @rynge do we actually still recommend that people use pool password? How does this work with the default GSI security that we ship in `osg-flock`?

Reviewer 2: Pool password is an alternative, we can be used if GSI is difficult to set up for the admins.

    CONDOR_HOST=$(FULL_HOSTNAME)
    UID_DOMAIN=$(FULL_HOSTNAME)
    FILESYSTEM_DOMAIN=$(FULL_HOSTNAME)
    
    #-- Authentication settings
    SEC_PASSWORD_FILE = /etc/condor/pool_password
    SEC_DEFAULT_AUTHENTICATION = REQUIRED
    SEC_DEFAULT_AUTHENTICATION_METHODS = FS,PASSWORD
    SEC_READ_AUTHENTICATION    = OPTIONAL
    SEC_CLIENT_AUTHENTICATION  = OPTIONAL
    SEC_ENABLE_MATCH_PASSWORD_AUTHENTICATION = TRUE
    DENY_WRITE         = anonymous@*
    DENY_ADMINISTRATOR = anonymous@*
    DENY_DAEMON        = anonymous@*
    DENY_NEGOTIATOR    = anonymous@*
    DENY_CLIENT        = anonymous@*
    
    #--  Privacy settings
    SEC_DEFAULT_ENCRYPTION = OPTIONAL
    SEC_DEFAULT_INTEGRITY = REQUIRED
    SEC_READ_INTEGRITY = OPTIONAL
    SEC_CLIENT_INTEGRITY = OPTIONAL
    SEC_READ_ENCRYPTION = OPTIONAL
    SEC_CLIENT_ENCRYPTION = OPTIONAL
    
    #-- With strong security, do not use IP based controls
    HOSTALLOW_WRITE = *
    
    # submitters/campus factories to flock to
    FLOCK_TO = flock.opensciencegrid.org
```    

Project Names
-------------

OSG will only run jobs tagged with a valid *ProjectName* - this is the main attribute

> used for accounting. New projects should be registered in the

Reviewer: ```suggestion
used for accounting.
```

> [OSG Topology Database](https://github.com/opensciencegrid/topology/blob/master/README.md).

Reviewer: ```suggestion
```

Please open at ticket to register a new project.
Jobs should specify which project to be accounted against by adding

> the *+ProjectNa,e* attribute. Note that the value is a string and hence

Reviewer: ```suggestion
the `+ProjectName` attribute. Note that the value is a string and hence
```

> the double quotes are required.  Example:

Reviewer: ```suggestion
the double quotes are required.  For example:
```

```file
+ProjectName = "Some_Name_Here"
```

> How to Get Help

Reviewer: To match the rest of our documentation

```suggestion
Getting Help
```
